### PR TITLE
introduced measurement module

### PR DESCRIPTION
**Pull request by pmelchior**
Methods that measure something off morphologies or sed or full data cubes in the model space are put into this new module. this clarifies where such methods go, and the methods don't use the full `component` as argument and then set implicit attributes. Instead, this is now done explicitly by the `update` methods.

### CODE REVIEW

**Comment by pmelchior** [comment]
As for the philosophy: it's not as much of a change as it is a clarification. `update` methods are, as you say, prox-like changes of contents of the components. they sometimes require measurements to do what they want to do. But we should distinguish those two tasks from each other. The measurement doesn't modify, that's why the don't get components as arguments.

File: scarlet/source.py [hunk 1]
```diff
@@ -319,6 +321,18 @@ def __init__(self, frame, sky_coord, observation, symmetric=True, monotonic=True
...
+        if self.symmetric:
+            if self.frame.psfs is None:
+                shape = (41, 41)
+                psf = generate_psf_image(gaussian, shape, amplitude=1, sigma=.9)
+                psf /= psf.max()
+                self._centroid_weight = psf
+            else:
+                self._centroid_weight = self.frame.psfs[0]
+
+        # ensure adherence to constraints
```
[the remark this answers is not in the document]
Comment (pmelchior) [reply scarlet/source.py]: That's exactly what the code does.

File: scarlet/source.py [hunk 2]
```diff
@@ -331,8 +345,10 @@ def update(self):
             it = 0
         else:
             it = self._parent.it
+
         # Update the central pixel location (pixel_center)
-        update.fit_pixel_center(self)
+        self.pixel_center = measurement.max_pixel(self.morph, self.pixel_center)
```
[the remark this answers is not in the document]
Comment (pmelchior) [reply scarlet/source.py]: As I wrote as a separate comment: updates change components, measurements do not. I think it is important that changes to components (like here the creation of `self.pixel_center` are not hidden in another function but clearly visible in the `update` method itself. Plus, we'll probably have a bunch of measurements that are independent of updates.

File: scarlet/source.py [hunk 3]
```diff
@@ -397,6 +414,17 @@ def __init__(self, frame, sky_coord, observation, bg_rms, thresh=1,
...
+        # create PSF as weight for centroid measurements
+        if self.symmetric:
+            if self.frame.psfs is None:
+                shape = (41, 41)
+                psf = generate_psf_image(gaussian, shape, amplitude=1, sigma=.9)
+                psf /= psf.max()
+                self._centroid_weight = psf
+            else:
+                self._centroid_weight = self.frame.psfs[0]
+
```
[the remark this answers is not in the document]
Comment (pmelchior) [reply scarlet/source.py]: ditto

File: scarlet/source.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fred3m) [reply scarlet/source.py]: Indeed. Perhaps I shouldn't review code that late at night...

**Comment by fred3m** [comment]
Ok, I guess I was a bit too tired when I looked at this last night. It's ok to merge.

File: scarlet/update.py [hunk 4]
```diff
@@ -5,40 +5,9 @@
...
-    Parameters
-    ----------
-    window: tuple of slices
-        Slices in y and x of the central region to include in the fit.
-        If `window` is `None` then only the 3x3 grid of pixels centered
-        on the previous center are used. If it is desired to use the entire
-        morphology just set `window=(slice(None), slice(None))`.
-    """
-    component.pixel_center = _fit_pixel_center(component.morph, component.pixel_center, window)
-    return component
```
[the remark this answers is not in the document]
Comment (pmelchior) [reply scarlet/update.py]: I thought we'd use `measurement` followed by the thing we want to measure, which is a bit of a break with terminology, where we normally use verbs for methods.

File: scarlet/measurement.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    if _morph.size < 500:
+        _bins = max(np.int(_morph.size/10), 1)
+        if _bins == 1:
+            return 0, _bins
+    hist, bins = np.histogram(np.log10(_morph).reshape(-1), _bins)
+    cutoff = np.where(hist == 0)[0]
+    # If all of the pixels are used there is no need to threshold
+    if len(cutoff) == 0:
+        return 0, _bins
+    return 10**bins[cutoff[-1]], _bins
```
[the remark this answers is not in the document]
Comment (pmelchior) [reply scarlet/measurement.py]: Line by line

**Pull request merged**